criu/cr-restore.c

```diff
@@ -3012,6 +3013,25 @@ static int prep_rseq(struct rst_rseq_param *rseq, ThreadCoreEntry *tc)
 	return 0;
 }
 
+#if defined(RSEQ_SIG)
+static void prep_libc_rseq_info(struct rst_rseq_param *rseq)
+{
+	if (!kdat.has_rseq) {
+		rseq->rseq_abi_pointer = 0;
+		return;
+	}
+
+	rseq->rseq_abi_pointer = encode_pointer(__criu_thread_pointer() + __rseq_offset);
```

Comment:
musl can have a different offset, can't it?

Reply:
This code is under `defined(RSEQ_SIG)` so it is used only for Glibc.

Reply:
upd: I've changed the condition to `#if defined(__GLIBC__) && defined(RSEQ_SIG)` just to be more explicit.

Reply:
upd2: added a comment with `TODO` about other libc libraries.

Reply:
It's unclear to me how the various scenarios are handled. You can have:

- glibc with rseq support, running on a Linux kernel without RSEQ support (either because the kernel is < 4.18, or because CONFIG_RSEQ=n),
- glibc with rseq support runnng on a Linux kernel with RSEQ support,
- glibc without rseq support running on kernel with RSEQ support,
- glibc without rseq support running on kernel without RSEQ support.

And you can build your application with recent kernel headers that have RSEQ support, but end up running on a Linux kernel without rseq support (either older kernel or CONFIG_RSEQ=n).

In librseq, I use "dlsym" to lookup whether glibc has the __rseq_offset and __rseq_size symbols to figure out at runtime if the glibc has rseq suport. I can use the __rseq_size value to figure out if glibc could successfully register rseq with the kernel. Ref. https://github.com/compudj/librseq/blob/master/src/rseq.c#L109

Reply:
Hi Mathieu,

thanks a lot for your attention to this topic.

Let me try to cover some of your points:
1. glibc with rseq support, running on the Linux kernel without RSEQ support. It's covered by check at `prep_libc_rseq_info` function. We just don't need to unregister anything and we can't restore processes with rseq dumped on the host, where rseq is not supported.
2. glibc with rseq support running on a Linux kernel with RSEQ support. covered by the same function. It finds `struct rseq` location and then it will be unregistered in the `restorer` blob.
3. glibc without rseq support running on the kernel with RSEQ support. It's okay. We just don't need to unregister rseq and can restore it without any problems.
4. glibc without rseq support running on kernel without RSEQ support. It's also okay. We will not try to unregister rseq and can't restore processes dumped on the system where rseq was supported.

>And you can build your application with recent kernel headers that have RSEQ support, but end up running on a Linux kernel without rseq support (either older kernel or CONFIG_RSEQ=n).

Good caught. But this particular case is not a problem because the check about rseq support in the kernel is dynamic (`kdat.has_rseq`). But I can analytically continue your idea and put a new question `what if criu was build on the system with old glibc (without rseq) and then run on the system with the newer glibc (with rseq)`. In this case we will fail because our check about glibc rseq support is compile-time:
```
#if defined(__GLIBC__) && defined(RSEQ_SIG)
static void prep_libc_rseq_info(struct rst_rseq_param *rseq)
{
	if (!kdat.has_rseq) {
		rseq->rseq_abi_pointer = 0;
		return;
	}

	rseq->rseq_abi_pointer = encode_pointer(__criu_thread_pointer() + __rseq_offset);
	rseq->rseq_abi_size = __rseq_size;
	rseq->signature = RSEQ_SIG;
}
#else
static void prep_libc_rseq_info(struct rst_rseq_param *rseq)
{
	/*
	 * TODO: handle built-in rseq on other libc'ies like musl
	 * We can do that using get_rseq_conf kernel feature.
	 *
	 * For now we just assume that other libc libraries are
	 * not registering rseq by default.
	 */
	rseq->rseq_abi_pointer = 0;
}
#endif
```

Fortunately, in most cases criu is built on the same userspace where it runs.

>In librseq, I use "dlsym" to lookup whether glibc has the __rseq_offset and __rseq_size symbols to figure out at runtime if the glibc has rseq suport. I can use the __rseq_size value to figure out if glibc could successfully register rseq with the kernel.

That's an interesting approach. I can try to take this way of rseq detection and unregistration.

cc @avagin

Reply:
We can continue discussion here https://github.com/checkpoint-restore/criu/issues/1814